src/jsCodegen/compileAstToJavaScript.ts

```diff
@@ -1,65 +1,104 @@
 import astHelper, { IAST } from '../parsing/astHelper';
 import { ReturnType } from '../parsing/convertXDMReturnValue';
+import { CodeGenContext } from './CodeGenContext';
 import { emitBaseExpr } from './emitBaseExpression';
 import {
 	acceptAst,
 	acceptFullyCompiledAst,
+	GeneratedCodeBaseType,
+	GeneratedCodeType,
+	getCompiledValueCode,
 	JavaScriptCompiledXPathResult,
 	PartialCompilationResult,
 	PartiallyCompiledAstAccepted,
 	rejectAst,
 } from './JavaScriptCompiledXPath';
-import { StaticContext } from './StaticContext';
 
 // Return all matching nodes.
-function emitEvaluationToNodes(identifier: string): PartiallyCompiledAstAccepted {
-	return acceptAst(`
+function emitEvaluationToNodes(
+	identifier: string,
+	generatedCodeType: GeneratedCodeType
+): PartiallyCompiledAstAccepted {
+	const [valueCode, valueCodeType] = getCompiledValueCode(identifier, generatedCodeType);
+	if (valueCodeType !== GeneratedCodeBaseType.Iterator) {
+		// throw new Error('Trying access generated code as an iterator while this is not the case.');
+		return acceptAst(`return [];`, { type: GeneratedCodeBaseType.Statement });
+	}
+
+	return acceptAst(
+		`
 	const nodes = [];
-	for (const node of ${identifier}(contextItem)) {
+	for (const node of ${valueCode}) {
 		nodes.push(node.value.node);
 	}
 	return nodes;
-	`);
+	`,
+		{ type: GeneratedCodeBaseType.Statement }
+	);
 }
 
 // Get effective boolean value.
-function emitEvaluationToBoolean(identifier: string): PartiallyCompiledAstAccepted {
-	return acceptAst(`
-	return determinePredicateTruthValue(${identifier}(contextItem));
-	`);
+function emitEvaluationToBoolean(
+	identifier: string,
+	generatedCodeType: GeneratedCodeType
+): PartiallyCompiledAstAccepted {
+	return acceptAst(
+		`return determinePredicateTruthValue(${
+			getCompiledValueCode(identifier, generatedCodeType)[0]
+		});`,
+		{ type: GeneratedCodeBaseType.Statement }
+	);
 }
 
 // Strings can just be returned as is so lets do that
-function emitEvaluationToString(identifier: string): PartiallyCompiledAstAccepted {
-	return acceptAst(`
-	return ${identifier}(contextItem);
-	`);
+function emitEvaluationToString(
+	identifier: string,
+	generatedCodeType: GeneratedCodeType
+): PartiallyCompiledAstAccepted {
+	return acceptAst(
+		`
+	return ${getCompiledValueCode(identifier, generatedCodeType)[0]};
+	`,
+		{ type: GeneratedCodeBaseType.Statement }
+	);
 }
 
-function emitEvaluationToFirstNode(identifier: string): PartiallyCompiledAstAccepted {
-	return acceptAst(`
-	const firstResult = ${identifier}(contextItem).next();
+function emitEvaluationToFirstNode(
+	identifier: string,
+	generatedCodeType: GeneratedCodeType
+): PartiallyCompiledAstAccepted {
+	const [valueCode, valueCodeType] = getCompiledValueCode(identifier, generatedCodeType);
+	if (valueCodeType !== GeneratedCodeBaseType.Iterator) {
+		throw new Error('Trying access generated code as an iterator while this is not the case.');
```

Comment:
For what kind of XPaths is this the case though?

Reply:
This was more meant as an assertion to make sure some other part of the code didn't pass it another GeneratedCodeType